field/TinyLang/Field/Generator.hs

```diff
@@ -228,81 +227,126 @@ arbitraryUnOpRing = elements [Neg]
+instance MonadState s m => MonadState s (GenT m) where
+    get = lift get
+    put = lift . put
+    state = lift . state
+
+
+type SGen f a = SGenT f Identity a
+
+-- Not sure if this is the right type
+runSGenT :: (Monad m) => Vars f -> SGenT f m a -> Gen (m a)
+runSGenT vars = fmap ((flip evalStateT) vars) . fmap (runSupplyT) . runGenT . unSGenT
+
+runSGen :: Vars f -> SGen f a -> Gen a
+runSGen vars = fmap runIdentity . runSGenT vars
+
+boundedArbitraryStmts :: forall m f. (Field f, Arbitrary f, MonadGen m, MonadSupply m, MonadState (Vars f) m)
+    => Int -> m (Statements f)
+boundedArbitraryStmts size =
+    mkStatements <$> frequency [ (1, pure [])
+                               , (10, arbStmts)
+                               ]
+    where
+        arbStmts = do
+            numStmts :: Int <- choose (1, size)
+            let perStmtSize = size `Prelude.div` numStmts
+            resize numStmts $ listOf $ boundedArbitraryStmt perStmtSize
+
+boundedArbitraryStmt :: forall m f. (Field f, Arbitrary f, MonadGen m, MonadSupply m, MonadState (Vars f) m)
+    => Int -> m (Statement f)
+boundedArbitraryStmt size
+    | size <= 1 = do
+          vars <- get
+          EAssert <$> boundedArbitraryExpr vars size
+    | otherwise = frequency stmtGens where
+          stmtGens = [ (3, withOneOfUnis $ \(_ :: Uni f a') -> do
+                               vars <- get
+                               uniVar <- genFreshUniVar @f @a'
+                               let vars' = Some uniVar : vars
+                                   size' = size - 1
+                               put vars'
+                               ELet uniVar <$> boundedArbitraryExpr vars' size')
+                     -- Generate a completely random assertion (unlikely to hold)
+                     , (1, do
+                               vars <- get
+                               EAssert <$> boundedArbitraryExpr vars size)
+                     -- generate a valid (but necessarily holding) contraint
+                     , (1, do
+                                let size' = size - 1
+                                vars <- get
+                                EAssert <$> boundedArbitraryComparisons vars size')
+                     -- generate an assertion of form @x binOp x@
+                     , (1, withOneOfBinAsserts $ \binOp -> do
+                               let size' = size `Prelude.div` 2
+                               vars <- get
+                               x <- boundedArbitraryExpr vars size'
+                               pure $ EAssert $ EAppBinOp binOp x x)
+                     -- generate a for loop
+                     , (3, do
+                               uniVar <- genFreshUniVar @f @(AField f)
+                               modify' (Some uniVar :)
+                               let size' = size - 1
+                                   start = arbitraryM
+                                   end   = arbitraryM
```

Comment:
We shouldn't generate loops with arbitrary bounds, that's too much to compute. Let's keep loops within, say, 10 iterations.